typing/errortrace.mli

```diff
@@ -106,28 +106,37 @@ val incompatible_fields :
 
 val swap_trace : ('a, 'variety) t -> ('a, 'variety) t
 
-(* The traces (['variety t]) are the core error types.  However, we bundle them
-   up into three "top-level" error types, which are used elsewhere:
-   [unification_error], [equality_error], and [moregen_error].  In the case of
-   [equality_error], this has to bundle in extra information; in general, it
-   distinguishes the three types of errors and allows us to distinguish traces
-   that are being built (or processed) from those that are complete and have
-   become the final error. *)
+(** The traces (['variety t]) are the core error types.  However, we bundle them
+    up into three "top-level" error types, which are used elsewhere:
+    [unification_error], [equality_error], and [moregen_error].  In the case of
+    [equality_error], this has to bundle in extra information; in general, it
+    distinguishes the three types of errors and allows us to distinguish traces
+    that are being built (or processed) from those that are complete and have
+    become the final error.  These error types have the invariants that their
+    traces are nonempty; we ensure that through three smart constructors with
+    matching names. *)
 
-type unification_error = { trace : unification error } [@@unboxed]
+type unification_error = private { trace : unification error } [@@unboxed]
```

Comment:
Some thoughts for that later day: I tried making the `error` type private, and that seemed, at least on a first cut, to be messier than making `unification_error` et al. private.  I very much like the idea of using a correct-by-construction nonempty type, although it may require some significant changes to `Ctype`, where `trace`s are often built up by raising `[]` in a function and then adding context.  I don't know if I'll have the time to get to making that change, but I would definitely support such a PR.